### ISSUE TITLE
`DataFrame.pivot` does not work with `index=None` even though function signature implies it is acceptable

### ISSUE THREAD

**Issue by henryharbeck** [pola-rs/polars#11592]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
df = pl.DataFrame(
    {
        "bar": ["y", "y", "y", "x"],
        "baz": [1, 2, 3, 4],
    }
)
df.pivot(values="baz", index=None, columns="bar", aggregate_function="sum")
```

### Log output

```shell
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_7501/2184148441.py in ?()
      4         "bar": ["y", "y", "y", "x"],
      5         "baz": [1, 2, 3, 4],
      6     }
      7 )
----> 8 df.pivot(values="baz", index=None, columns="bar", aggregate_function="sum")

~/development/polars_help/venv/lib/python3.11/site-packages/polars/dataframe/frame.py in ?(self, values, index, columns, aggregate_function, maintain_order, sort_columns, separator)
   7024         else:
   7025             aggregate_expr = aggregate_function._pyexpr
   7026 
   7027         return self._from_pydf(
-> 7028             self._df.pivot_expr(
   7029                 values,
   7030                 index,
   7031                 columns,

TypeError: argument 'index': 'NoneType' object cannot be converted to 'PyString'
```

### Issue description

`DataFrame.pivot` does not work with `index=None`.

The function signature implies it is acceptable by type hinting None as an option.

However, the docstring says 
> index - One or multiple keys to group by.

potentially implying that `None` is not valid as that would be grouping by 0 keys.

### Expected behavior

In my opinion, there is no reason `index=None` should not be valid. 
It would just mean that the output of the `pivot` would always be a single row.

For the example provided, the expected output would be

```
┌─────┬─────┐
│ y   ┆ x   │
│ --- ┆ --- │
│ i64 ┆ i64 │
╞═════╪═════╡
│ 6   ┆ 4   │
└─────┴─────┘
```

The docstring for the `index` parameter should also be updated to be clear that passing `None` is valid - or at least not imply that it is invalid. 

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.7
Index type:          UInt32
Platform:            Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.31
Python:              3.11.4 (main, Jun  8 2023, 17:02:11) [GCC 9.4.0]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              <not installed>
gevent:              <not installed>
matplotlib:          <not installed>
numpy:               <not installed>
openpyxl:            <not installed>
pandas:              <not installed>
pyarrow:             <not installed>
pydantic:            <not installed>
pyiceberg:           <not installed>
pyxlsb:              <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

</details>

**Comment by deanm0000** [comment]
What you're intending to get back is the transpose of group_by.agg

```
df.group_by('bar').agg(pl.col('baz')).transpose()
```

**Comment by henryharbeck** [comment]
@deanm0000, the transpose does not provide header names like pivot does. Obviously can promote the first row as headers afterwards though.

A perhaps simpler workaround for the issue would be to create a literal column (with a single unique value), use that as the `index` to pivot and then drop it after.

```python
(
    df.with_columns(pl.lit(1))
    .pivot(values="baz", index="literal", columns="bar", aggregate_function="sum")
    .drop("literal")
)
```
The above produces the expected output. 

Workarounds or alternative approaches aside, I created this issue as I believe (and the type hints in the function indicate) that `index=None` should be valid, but is currently raising an error.

**Comment by deanm0000** [comment]
I wasn't trying to discount the request, just trying to help. That said, what I put in earlier was on mobile without looking at how it worked. A more complete version would be

```
(
    df.group_by('bar', maintain_order=True).sum()
    .select('baz').transpose(column_names=df['bar'].unique())
    )
```

A way that doesn't use transpose and so might be more efficient...
```
(
    df
    .group_by('bar', maintain_order=True)
    .agg(pl.col('baz').sum())
    .select(pl.col('baz').implode().list.to_struct().struct.rename_fields(df['bar'].unique()))
    .unnest('baz')
)
```

**Comment by gab23r** [comment]
This bug reminds me of this one: #10075

**Comment by henryharbeck** [comment]
Hey @MarcoGorelli, as you look to have been in the world of `pivot` lately, thought you may appreciate a heads up on this one.
I have checked again on 0.20.7, and this still produces the same error.

And on `main`, there is still the discrepancy for `index` between the type hint (`None` as an option) and docstring ("One or multiple keys to group by.", implying that `None` is not okay.).

**Comment by MarcoGorelli** [comment]
Hey

Just looking at this again, and I'm not sure about adding complexity to `pivot`, it's already pretty complicated

The desired functionality can be achieved with
```
df.group_by('bar').agg(pl.sum('baz')).transpose(column_names='bar')

shape: (1, 2)
┌─────┬─────┐
│ x   ┆ y   │
│ --- ┆ --- │
│ i64 ┆ i64 │
╞═════╪═════╡
│ 4   ┆ 6   │
└─────┴─────┘
```
which is a bit expensive, but then again so is `pivot`. And it's arguably clearer than
```
df.pivot(values="baz", index=None, columns="bar", aggregate_function="sum")
```
?

Something else I feel uneasy about is this:
- `values=None` means "use all remaining columns if `index` and `columns` have already been specified"

So, for consistency, I'd expect `index=None` to mean "use all remaining columns if `columns` and `values` have been specified". I have some recollection that this is what @mcrumiller was going for at some point, rather than the 1-row variant suggested here

**Comment by ohines** [comment]
@MarcoGorelli
It seems there are two points:
* Can the user specify that no index cols should be used and a single output row be produced e.g. `index=[]`.
* should `index=None` correspond to `index=[]` or 'all remaining columns'.

I think that the second point is distinct from the first. Perhaps we could allow e.g.
```python
df = pl.DataFrame(
    {
        "foo": ["A", "B", "C"],
        "N": [1, 2, 3],
        "M": [4, 5, 6],
    }
)
df.pivot(index=[], columns="foo", values=None, aggregate_function=None)

shape: (1, 6)
┌─────────┬─────────┬─────────┬─────────┬─────────┬─────────┐
│ N_foo_A ┆ N_foo_B ┆ N_foo_C ┆ M_foo_A ┆ M_foo_B ┆ M_foo_C │
│ ---     ┆ ---     ┆ ---     ┆ ---     ┆ ---     ┆ ---     │
│ i64     ┆ i64     ┆ i64     ┆ i64     ┆ i64     ┆ i64     │
╞═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ 1       ┆ 2       ┆ 3       ┆ 4       ┆ 5       ┆ 6       │
└─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘

```
Note that this behaviour is already implemented in #15855
so that PR would just have to be modified to remove the `None` becomes `[]` logic.
